src/tracker/client.rs

```diff
@@ -0,0 +1,320 @@
+      downloaded: 0x0000,
+      left: u64::MAX,
+      uploaded: 0x0000,
+      event: 0x0000,
+      ip_address: 0x0000,
+      num_want: u32::MAX,
+      port: self.sock.local_addr().context(error::Network)?.port(),
+    };
+    let mut buf = [0u8; 1024];
+    let (_, payload) = self.exchange(&req, &mut buf)?;
+
+    Client::parse_compact_peer_list(self.is_ipv6, payload)
+  }
+
+  fn exchange<T: Request>(&self, req: &T, rxbuf: &'a mut [u8]) -> Result<(T::Response, &'a [u8])> {
+    let msg = req.serialize();
+    let mut len_read: usize = 0;
+
+    for _ in 0..3 {
+      self.sock.send(&msg).context(error::Network)?;
+      if let Ok(len) = self.sock.recv(rxbuf) {
+        len_read = len;
+        break;
+      }
+    }
+
+    if len_read == 0 {
+      return Err(Error::TrackerTimeout {
+        host_port: self.host_port.clone(),
+      });
+    }
+
+    let (resp, payload) = T::Response::deserialize(&rxbuf[..len_read])?;
+    if resp.transaction_id() != req.transaction_id() || resp.action() != req.action() {
+      return Err(Error::TrackerResponse);
+    }
+
+    Ok((resp, payload))
+  }
+
+  fn parse_compact_peer_list(is_ipv6: bool, buf: &[u8]) -> Result<Vec<SocketAddr>> {
+    let mut subswarm = Vec::<SocketAddr>::new();
+    let stride = if is_ipv6 { 18 } else { 6 };
+
+    let chunks = buf.chunks_exact(stride);
+    if !chunks.remainder().is_empty() {
+      return Err(Error::TrackerResponse);
+    }
+
+    for hostpost in chunks {
+      let (ip, port) = hostpost.split_at(stride - 2);
+      let ip = if is_ipv6 {
+        let buf: [u8; 16] = ip[0..16]
+          .try_into()
+          .invariant_unwrap("iterator guarantees bounds are OK");
+        IpAddr::from(std::net::Ipv6Addr::from(buf))
+      } else {
+        IpAddr::from(std::net::Ipv4Addr::new(ip[0], ip[1], ip[2], ip[3]))
+      };
+
+      let port = u16::from_be_bytes(
+        port
+          .try_into()
+          .invariant_unwrap("iterator guarantees bounds are OK"),
+      );
+
+      subswarm.push((ip, port).into());
+    }
+
+    Ok(subswarm)
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use action::Action;
+  use std::{thread, time::Duration};
+
+  #[cfg(test)]
+  pub(crate) fn new_dummy_metainfo() -> Metainfo {
+    Metainfo {
+      announce: None,
+      announce_list: None,
+      nodes: None,
+      comment: None,
+      created_by: None,
+      creation_date: None,
+      encoding: None,
+      info: Info {
+        private: None,
+        piece_length: Bytes(16 * 1024),
+        source: None,
+        name: "testing".into(),
+        pieces: PieceList::from_pieces(["test", "data"]),
+        mode: Mode::Single {
+          length: Bytes(2 * 16 * 1024),
+          md5sum: None,
+        },
+        update_url: None,
+      },
+    }
+  }
+
+  fn simulate_tracker_announce_response_helper(server: UdpSocket, targets: Vec<u8>, slowly: bool) {
+    let mut rng = rand::thread_rng();
+    let mut buf = [0u8; 8192];
+
+    // connect exchange
+    {
```

Comment:
I've scoped this and the next one because clippy was complaining about the shadowing.